java/ql/lib/semmle/code/java/security/CsrfUnprotectedRequestTypeQuery.qll

```diff
@@ -0,0 +1,256 @@
+    )
+  }
+}
+
+/** A method declared in `java.sql.PreparedStatement` that updates a database. */
+private class PreparedStatementDatabaseUpdateMethod extends DatabaseUpdateMethod {
+  PreparedStatementDatabaseUpdateMethod() {
+    this instanceof PreparedStatementExecuteUpdateMethod or
+    this instanceof PreparedStatementExecuteLargeUpdateMethod
+  }
+}
+
+/** A method found via the sql-injection sink models which may update a database. */
+private class SqlInjectionDatabaseUpdateMethod extends DatabaseUpdateMethod {
+  SqlInjectionDatabaseUpdateMethod() {
+    exists(DataFlow::Node n | this = n.asExpr().(Argument).getCall().getCallee() |
+      sinkNode(n, "sql-injection") and
+      // do not include `executeQuery` since it is typically used with a select statement
+      this.hasName([
+          "delete", "insert", "update", "batchUpdate", "executeUpdate", "executeLargeUpdate",
+          "execute"
+        ])
+    )
+  }
+}
+
+/**
+ * A taint-tracking configuration for reasoning about SQL statements that update
+ * a database via a call to an `execute` method.
+ */
+private module SqlExecuteConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    exists(StringLiteral sl | source.asExpr() = sl |
+      sl.getValue().regexpMatch("^(?i)(insert|update|delete).*")
+    )
+  }
+
+  predicate isSink(DataFlow::Node sink) {
+    exists(Method m | m = sink.asExpr().(Argument).getCall().getCallee() |
+      m instanceof SqlInjectionDatabaseUpdateMethod and
+      m.hasName("execute")
+    )
+  }
+}
+
+/**
+ * Tracks flow from SQL statements that update a database to the argument of
+ * an `execute` method call.
+ */
+private module SqlExecuteFlow = TaintTracking::Global<SqlExecuteConfig>;
+
+/** Provides classes and predicates representing call graph paths. */
+module CallGraph {
+  private newtype TCallPathNode =
+    TMethod(Method m) or
+    TCall(Call c)
+
+  /** A node in a call path graph */
+  class CallPathNode extends TCallPathNode {
+    /** Gets the method corresponding to this `CallPathNode`, if any. */
+    Method asMethod() { this = TMethod(result) }
+
+    /** Gets the call corresponding to this `CallPathNode`, if any. */
+    Call asCall() { this = TCall(result) }
+
+    /** Gets the string representation of this `CallPathNode`. */
+    string toString() {
+      result = this.asMethod().toString()
+      or
+      result = this.asCall().toString()
+    }
+
+    private CallPathNode getACallee() {
+      [viableCallable(this.asCall()), this.asCall().getCallee()] = result.asMethod()
+    }
+
+    /** Gets a successor node of this `CallPathNode`, if any. */
+    CallPathNode getASuccessor() {
+      this.asMethod() = result.asCall().getEnclosingCallable()
+      or
+      result = this.getACallee() and
+      (
+        exists(CallPathNode p |
+          p = this.getACallee() and
+          p.asMethod() instanceof DatabaseUpdateMethod
+        )
```

Comment:
Factor this snippet out into a private nomagic'ed member predicate (e.g. called `canTargetDatabaseUpdateMethod()`) to avoid any potential cartesian products between callee sets. Notice that the current QL can potentially join `result = this.getACallee()` with `p = this.getACallee()`, which is a needless blowup.

Reply:
Updated in [642a2b1](https://github.com/github/codeql/pull/18288/commits/642a2b14518c18dff7ef97ea0b70aa7e26e69688)